pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -678,7 +678,12 @@ func (l *LibvirtDomainManager) generateCloudInitEmptyISO(vmi *v1.VirtualMachineI
 //
 // The Domain.Spec can be alterned in this function and any changes
 // made to the domain will get set in libvirt after this function exits.
-func (l *LibvirtDomainManager) preStartHook(vmi *v1.VirtualMachineInstance, domain *api.Domain, generateEmptyIsos bool, options *cmdv1.VirtualMachineOptions) (*api.Domain, error) {
+func (l *LibvirtDomainManager) preStartHook(
```

Comment:
You may be right that the method mutates the inputted domain, but it is very hard to be sure really.
From a caller perspective, before this change, I understood that it is suppose to change as it returned the object, but now a common reader would be surprised.

The problem is that it does mutate it and I would really prefer to fix that instead of making it hidden.

On the other hand, you could claim that you just express the reality here and it is better to see the problem over hiding it. I do not know what is better. :/

Reply:
> On the other hand, you could claim that you just express the reality here and it is better to see the problem over hiding it. I do not know what is better. :/

I'm fine either way, but we should come up with a clear way to define when input is being changed (e.g: documentation, variable prefix/suffix).

If we think about functions like `append`, because the slice might be nil, it'll do the allocation and return the pointer to the new slice. The same should happen when slice's `cap` is reached and `append` will allocate a new bigger chunk of memory.

I don't think that would ever happen for most functions that will do changes `api.DomainSpec`, that is, the initial pointer will be kept unchanged. For that reason, I think it is somewhat safe to do what I did. `preStartHook` will do what it advertises and any related changes will be propagated to internals of `api.DomainSpec`.

Again, either way is fine for me, I'm just in favour of  `team consistency` overall ;)